vars/runATH.groovy

```diff
@@ -0,0 +1,156 @@
+                                                             userRemoteConfigs: [[url: "${athUrl}"]]]
+            }
+        }
+        dir(athSourcesFolder) {
+            def uid = sh(script: "id -u", returnStdout: true)
+            def gid = sh(script: "id -g", returnStdout: true)
+            athContainerImage = docker.build('jenkins/ath', "--build-arg=uid='$uid' --build-arg=gid='$gid' -f src/main/resources/ath-container/Dockerfile .")
+            // We may need to run things in parallel later, so avoid several checkouts
+            stash name: "athSources"
+        }
+
+        // Jenkins war
+        sh("curl -o jenkins.war -L ${jenkinsURl}")
+        stash includes: '*.war', name: 'jenkinsWar'
+
+    }
+    // Workaround for https://issues.jenkins-ci.org/browse/JENKINS-27092
+    if (shouldStop) {
+
+        return
+    }
+    // End of workaround
+
+    stage("Running ATH") {
+        def testsToRun = metadata.tests?.join(",")
+        def categoriesToRun = metadata.categories?.join(",")
+        def browsers = metadata.browsers ?: ["firefox"]
+        def failFast = metadata.failFast ?: false
+        def rerunCount = metadata.rerunFailingTestsCount ?: 0
+        def localSnapshots = metadata.useLocalSnapshots ?: true
+
+        if (testsToRun == null && categoriesToRun == null) {
+            categoriesToRun = defaultCategory
+        }
+
+        def testingbranches = ["failFast": failFast]
+        for (browser in browsers) {
+            // TODO remove this if else block once other browser are supported by infra
+            if (browser == "firefox") {
+
+                def currentBrowser = browser
+                def containerArgs = "-v /var/run/docker.sock:/var/run/docker.sock -e LOCAL_SNAPSHOTS=${localSnapshots} -e SHARED_DOCKER_SERVICE=true -u ath-user"
+                def commandBase = "./run.sh ${currentBrowser} jenkins.war -Dmaven.test.failure.ignore=true -DforkCount=1 -B -Dsurefire.rerunFailingTestsCount=${rerunCount}"
+
+                if (testsToRun) {
+                    testingbranches["ATH individual tests-${currentBrowser}"] = {
+                        dir("test${currentBrowser}") {
+                            unstash name: "athSources"
+                            unstash name: "jenkinsWar"
+                            def command = commandBase + " -Dtest=${testsToRun}"
+                            athContainerImage.inside(containerArgs) {
+                                realtimeJUnit(testResults: 'target/surefire-reports/TEST-*.xml', testDataPublishers: [[$class: 'AttachmentPublisher']]) {
+                                    sh 'eval "$(./vnc.sh)" && ' + command
+                                }
+
+                            }
+                        }
+                    }
+                }
+                if (categoriesToRun) {
+                    testingbranches["ATH categories-${currentBrowser}"] = {
```

Comment:
BTW see https://github.com/jenkinsci/parallel-test-executor-plugin/pull/34 and https://github.com/jenkinsci/acceptance-test-harness/pull/380. Not sure it matters for the use case here, which presumably involves only a small number of tests per build.

Reply:
> Not sure it matters for the use case here, which presumably involves only a small number of tests per build.

Yeah, that is my understanding and the reason why I did not use parallel test executor plugin